### PR TITLE
Code coverage for JsonWriterHelper JsonWriter

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 1]
```diff
@@ -29,6 +31,24 @@ public static void WriteNumber(bool indented)
...
+        {
+            var buffer = new ArrayBufferWriter<byte>(1024);
+            using (JsonDocument doc = JsonDocument.Parse($" [ {value} ]", s_options))
+            {
+                JsonElement target = doc.RootElement[0];
+
+                var writer = new Utf8JsonWriter(buffer);
+
+                target.WriteValue(writer);
+            }
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Seems like an AssertContents would be useful

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 2]
```diff
@@ -437,6 +457,121 @@ public static void WriteNumberScientificAsProperty(bool indented)
                 "{\"m\\u00eal\\u00e9e\":1e6}");
         }
 
+        [Fact]
+        public static void WriteValueUnicodeEscapeSequence()
+        {
+            char[] chars = new char[] { (char)0x0421, (char)0x0430, (char)0x0439, (char)0x043D, ' ', (char)0x0443, (char)0x0443, '?', (char)0x04E9, (char)0x04E9 };
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Some of these lines are longer than github's default code view width (requiring horizontal scrolling).

Please break them up to improve the legibility of this code on github.

Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: This array gets passed to `new string(chars)` twice.  Maybe just defining it as

```C#
string input = "\u0421\u0430\u0439\u043D \u0443?\u04E9\u04E9";
```

would be better?

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 3]
```diff
@@ -437,6 +457,121 @@ public static void WriteNumberScientificAsProperty(bool indented)
...
+                var writer = new Utf8JsonWriter(buffer);
+                target.WriteValue(writer);
+
+                Assert.Throws<ArgumentException>(() => target.WriteValue(writer));
+            }
+        }
+
+        private static string GetEscapedExpectedString(bool prettyPrint, string value, StringEscapeHandling escaping)
+        {
+            var ms = new MemoryStream();
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Please put the MemoryStream in a `using` block.

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 4]
```diff
@@ -29,6 +31,24 @@ public static void WriteNumber(bool indented)
...
+        //        [InlineData("1.2e+13")] Enable after https://github.com/dotnet/corefx/issues/39139 resolved
+        //        [InlineData("1.2E-3")]
+        public static void WriteScientificNumber(string value)
+        {
+            var buffer = new ArrayBufferWriter<byte>(1024);
+            using (JsonDocument doc = JsonDocument.Parse($" [ {value} ]", s_options))
+            {
+                JsonElement target = doc.RootElement[0];
+
+                var writer = new Utf8JsonWriter(buffer);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Utf8JsonWriter is IDisposable, so all of these writers should be in `using` blocks

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: ok

Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: yeah

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 5]
```diff
@@ -437,6 +449,121 @@ public static void WriteNumberScientificAsProperty(bool indented)
...
+            using (JsonDocument doc = JsonDocument.Parse(sequence, s_options))
+            {
+                JsonElement target = doc.RootElement[0];
+
+                using (var writer = new Utf8JsonWriter(buffer))
+                {
+                    target.WriteValue(writer);
+                    writer.Flush();
+                }
+                string str = Encoding.UTF8.GetString(buffer.WrittenSpan);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: You don't use this variable, and this line is making the netfx build fail:

```
JsonElementWriteTests.cs(470,54): error CS1503: Argument 1: cannot convert from 'System.ReadOnlySpan<byte>' to 'byte[]' [D:\a\1\s\src\System.Text.Json\tests\System.Text.Json.Tests.csproj]
##[error]JsonElementWriteTests.cs(470,54): error CS1503: (NETCORE_ENGINEERING_TELEMETRY=Build) Argument 1: cannot convert from 'System.ReadOnlySpan<byte>' to 'byte[]'
```

Seems like the line can just be deleted.

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 6]
```diff
@@ -29,6 +31,16 @@ public static void WriteNumber(bool indented)
             WriteSimpleValue(indented, "42");
         }
 
+        [Theory]
+        [InlineData("12E-3")]
+        [InlineData("12e+3")]
+        //        [InlineData("1.2e+13")] Enable after https://github.com/dotnet/corefx/issues/39139 resolved
+        //        [InlineData("1.2E-3")]
+        public static void WriteScientificNumber(string value)
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: 39139 is resolved now, but I think that this method is redundant with tests added via the PR (#39406).  I'm guessing there's a merge conflict at this point now, and the easiest resolution is probably to just remove this method.

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 5 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: uh, forgot to remove

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 6 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: checked the PR, it didn't covered test for capital E, i think will keep the test

Comment (bartonjs) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: Can you merge these cases into the existing theory, then?

Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: sure

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 7]
```diff
@@ -460,6 +464,122 @@ public static void WriteNumberScientificAsProperty(bool indented)
...
+        [Fact]
+        public static void WriteValueUnicodeEscapeSequence()
+        {
+            string unicodeString = "\u0421\u0430\u0439\u043D \u0443\u0443? \u04E9\u04E9";
+            string json = $"[\"{unicodeString}\"]";
+            var buffer = new ArrayBufferWriter<byte>(1024);
+            ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(Encoding.UTF8.GetBytes(json), 1);
+            string expectedStr = GetEscapedExpectedString(unicodeString, StringEscapeHandling.EscapeNonAscii);
+
+            using (JsonDocument doc = JsonDocument.Parse(sequence, s_options))
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: Many of these tests seem like they'd be more appropriate at the writer level, because they're not using anything specific from JsonDocument.  Am I missing JsonDocument's involvement?

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 8]
```diff
@@ -460,6 +464,122 @@ public static void WriteNumberScientificAsProperty(bool indented)
...
+            var buffer = new ArrayBufferWriter<byte>(1024);
+            byte[] temp = Encoding.UTF8.GetBytes(json);
+            ReadOnlySequence<byte> sequence = JsonTestHelper.GetSequence(temp, 1);
+            string expectedStr = GetEscapedExpectedString(unicodeString, StringEscapeHandling.EscapeNonAscii);
+
+            using (JsonDocument doc = JsonDocument.Parse(sequence, s_options))
+            {
+                JsonElement target = doc.RootElement[0];
+
+                var writer = new Utf8JsonWriter(buffer);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: It looks like only some of the new tests put writers are in using statements; is that intentional?

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 7 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: Not sure what you mean exactly. It is getting JsonElement from JsonDocument, the test is for JsonElement.WriteTo not for Utf8JsonWriter

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 8 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: I see everywhere were else it is used without using, following existing pattern

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 7 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: I understand that the test _is_ using JsonDocument/JsonElement, but the question is "why?"  What's the scenario that's being tested that requires using JsonDocument/JsonElement.  The method name feels like it's testing something in the writer, which means that the test could be written without using JsonDocument/JsonElement at all.

It's using sequences, which aren't really interesting to JsonDocument, because JsonDocument.Parse just drains all the sequence contents into one rented array; so the write isn't doing anything clever.

I'm really just trying to understand what this test is testing... what were we not already testing that we are now?

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 8 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: Please add the using to all new Utf8JsonWriter instances.

Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: sure

File: src/System.Text.Json/tests/JsonElementWriteTests.cs (hunk 7 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonElementWriteTests.cs]: That might explain why I was not able to hit error nous conditions wanted to cover with whatever bad data used...

I just trying to improve coverage for JsonWriterHelper.Escaping.cs, I added this because it was calling https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs#L234 for which I was adding coverage. But I see that reference is removed just 3 hours ago …

As you know I don't really familiar with underlying code design just following what is calling the those section or what tests running around that. But seems the code i was trying to cove was invalid.

File: src/System.Text.Json/tests/JsonElementWriteTests.cs [hunk 9]
```diff
@@ -523,6 +526,43 @@ public static void WriteNumberScientificAsProperty(bool indented)
...
+                using (var writer = new Utf8JsonWriter(buffer))
+                {
+                    target.WriteTo(writer);
+                    writer.Flush();
+                }
+                AssertContents(expectedStr, buffer);
+            }
+        }
+
+        private static string GetEscapedExpectedString(string value, StringEscapeHandling escaping)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonElementWriteTests.cs]: nit: You could use a `StringWriter` instead, and skip the transcoding to utf-8 and back. But this is outside of the scope of this PR and if we are changing it, we may want to fix it in the existing writer tests as well:
https://github.com/dotnet/corefx/blob/05c0f516cf20cd6ae108ba034984b0cdc740732e/src/System.Text.Json/tests/Utf8JsonWriterTests.cs#L5383-L5904

```C#
private static string GetEscapedExpectedString(string value, StringEscapeHandling escaping)
{
    using (TextWriter stringWriter = new StringWriter())
    using (var json = new JsonTextWriter(stringWriter)
    {
        StringEscapeHandling = escaping
    })
    {
        json.WriteValue(value);
        json.Flush();
        return stringWriter.ToString();
    }
}
```

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs [hunk 10]
```diff
@@ -3097,6 +3148,17 @@ public void EscapeSurrogatePairs(bool formatted, bool skipValidation)
             }
         }
 
+        [Fact]
+        public void EscapeSurrogatePairsThrowsHighSurrogateMissing()
+        {
+            string propertyName = "a \uD800 \uDC00\uDE6D a";
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Why are there space character in here? The test name says "High Surrogate Missing", but it looks like the writer will fail with low surrogate missing because it will try to write these two characters: \uD8000, SPACE, and the surrogate pair ends up missing the low surrogate following it (\uDC00).

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs [hunk 11]
```diff
@@ -2311,6 +2311,57 @@ public void WriteBase64StringEscaped(bool formatted, bool skipValidation)
             Assert.True(Encoding.UTF8.GetBytes(propertyName).AsSpan().SequenceEqual(propertyNameSpanUtf8));
         }
 
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void WriteBase64StringEscapeLengthBigger(bool formatted, bool skipValidation)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: nit: **If feasible,** consider instead combining this test case with the ones above, into a single `WriteBase64String` test, similar to `WriteDateTimesValue`. So, instead of having 3 separate tests that primarily just defer on the property name, we have 1 test, with different inline data.

https://github.com/dotnet/corefx/blob/063e38c7bc66e076fdcc7521e4c69650baae9427/src/System.Text.Json/tests/Utf8JsonWriterTests.cs#L2206-L2312

https://github.com/dotnet/corefx/blob/063e38c7bc66e076fdcc7521e4c69650baae9427/src/System.Text.Json/tests/Utf8JsonWriterTests.cs#L4255-L4268

Comment (buyaa-n) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Isn't it better to have separate methods which names illustrates what exactly testing, I would prefer to move shared/similar code section into a method but leave separate test methods

Comment (buyaa-n) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Combined tests instead of keeping separate, updated string values to illustrate what it is testing

**Comment by ahsonkhan** [comment]
Most of the remaining JSON issues (and PRs like this one) fall in the "test enhancement" or "documentation" bucket which are intended for 3.0 and still meet the 3.0-preview8 bug bar. The intention is to get these in for 3.0. Hence, modifying the milestone back to 3.0. @karelz, please let me know if you disagree or had some other reason to move the milestone to 5.0.

cc @ericstj

**Comment by ahsonkhan** [comment]
@buyaa-n helped with this and the PR is up: https://github.com/dotnet/corefx/pull/39612
